File: semestr-8/Jezyk C#/alexander/Csharp/Archiwum/2019/C-ID06TC1/MergeThreadTest/MergeThreadTest/Merge.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace MergeThreadTest
{
    class Merge
    {
        private int[] Tb, Tmp;
        class LR
        {
            public int L { get; set; }
            public int R { get; set; }
            public LR(int L,int R)
            {
                this.L = L;
                this.R = R;
            }
        }
        private Merge(int[] Tb)
        {
            this.Tb = Tb;
            Tmp=new int[Tb.Length];
            MakeThread(new LR(0, Tb.Length)).Join();
        }
        private void ThreadRun(LR lr)
        {
            int L = lr.L, R = lr.R;
            Thread tha = null, thb = null;
            int M = (L + R) >> 1;
            if (M - L > 1) tha = MakeThread(new LR(L, M));
            if (R - M > 1) thb = MakeThread(new LR(M, R));
            if (tha != null) tha.Join();
            if (thb != null) thb.Join();
            int a = L, b = M, p = L;
            while ((a < M) && (b < R))
            {
                if (Tb[a] < Tb[b]) Tmp[p++] = Tb[a++];
                else Tmp[p++] = Tb[b++];
            }
            while (a < M) Tmp[p++] = Tb[a++];
            while (b < R) Tmp[p++] = Tb[b++];
            Array.Copy(Tmp, L, Tb, L, R - L);
        }
        private void run(object obj)
        {
            ThreadRun((LR)obj);
        }
        private Thread MakeThread(LR lr)
        {
            Thread th = new Thread(run);
            th.Start(lr);
            return th;
        }
        public static int[] Sort(int[] Src)
        {
            return new Merge(Src).Tb;
        }
    }
}

File: semestr-8/Programowanie obiektowe w Java/alexander/POJ/Archiwum/2019w/ID07IO1/MergeSortTest/src/mergesorttest/Merge.java
package mergesorttest;

public class Merge 
{
    private final int[] tab;
    private final int[] add;
    public static void sort(int[] tab) { new Merge(tab); }
    private Merge(int[] tab)
    {
        this.tab=tab;
        add=new int[tab.length];
        new MergeTh(0,tab.length).safejoin();
    }
    private class MergeTh extends Thread
    {
        private int aBegin,bEnd;
        public MergeTh(int aBegin,int bEnd)
        {
            this.aBegin=aBegin;
            this.bEnd=bEnd;
            start();
        }
        public void safejoin()
        {
            try { join(); } catch (InterruptedException ex) {}
        }
        @Override public void run()
        {
        }
    }
}
